autoload/phpactor/project/assigner.vim

```diff
@@ -0,0 +1,118 @@
+function! phpactor#project#assigner#assignProjectToBuffer(filename)
+  if !exists('g:phpactorLoaded')
+    return
+  endif
+
+  if exists('b:project')
+    return
+  endif
+  let l:project = g:phpactorBufferMatcher.resolveProjectForFile(a:filename)
+  if v:null is l:project
+    return
+  endif
+  let b:project = l:project
+  if g:phpactorBufferMatcher.repository.hasProject(b:project)
+    return
+  endif
+
+  call g:phpactorBufferMatcher.repository.addProject(b:project)
+endfunction
+
+function! phpactor#project#assigner#create(projectRepository, projectRootMarkers, filesystemRootMarkers, initialCwd) abort
+  let l:filesystemRootMarkers = a:filesystemRootMarkers
+  if index(l:filesystemRootMarkers, '/') < 0
+    call add(l:filesystemRootMarkers, '/')
+  endif
+
+  return {
+        \ 'initialCwd': a:initialCwd,
+        \ 'projectRootMarkers': a:projectRootMarkers,
+        \ 'filesystemRootMarkers': l:filesystemRootMarkers,
+        \ 'repository': a:projectRepository,
+        \ 'resolveProjectForFile': function('s:resolveProjectForFile')
+        \ }
+endfunction
+
+function s:resolveProjectForFile(file) dict abort
+  let l:project = self.repository.findProjectContainingFile(a:file)
+
+  if type(l:project) == type({})
+    return l:project
+  endif
+
+  " early version to refactor or change
+  " heredoc are available since vim 8.1.1354
+  let l:message = printf(join([
+        \ 'Phpactor [RPC]',
+        \ 'There is no project enabled for "%s" yet.',
+        \ 'Select the way of assign the project root',
+        \ 'If cancelled, "%s" will be used.'
+        \], "\n"), a:file, self.initialCwd)
+
+  " see phpactor#_path
+  let l:initialDirectory = fnamemodify(a:file, ':p:h')
+  let l:rootDirByMarker = s:searchDirectoryUpwardForRootMarkers(
+        \ l:initialDirectory,
+        \ self.projectRootMarkers,
+        \ self.filesystemRootMarkers
+        \ )
+
+  let l:choices = [
+        \ {
+        \ 'action': { -> self.initialCwd },
+        \ 'message': l:message
+        \ },
+        \ {
+        \ 'action': function('input', ['Enter file path: ', l:initialDirectory, 'file']),
+        \ 'message': printf('manual (default "%s")', l:initialDirectory)
+        \ }
+        \ ]
+
+  if v:null != l:rootDirByMarker
+    let l:item = {
+          \ 'action': { -> l:rootDirByMarker },
+          \ 'message': 'autodetected by root marker: '.l:rootDirByMarker
+          \ }
+    call add(l:choices, l:item)
+  endif
+
+  let l:choice = v:null
+  while index(range(0, len(l:choices)-1), l:choice) < 0
```

Comment:
```suggestion
  while index(range(1, len(l:choices)-1), l:choice) < 0
```
I would ignore `0` since it's the prompt I feel like it should not be a valid choice

Reply:
Did you try it along with `0` as the start value (which is btw the good idea)? After press Enter (default value) `l:choices` becomes `0` and the loop requires to be break.

Reply:
Behaviour after pressing `Ctrl-C` requires more attention. In the current version it causes 
```
"/tmp/x2.php" [New File]                                                                                                                                                                                                                        
Phpactor [RPC]                                                                                                                                                                                                                                  
There is no project enabled for "/tmp/x2.php" yet.                                                                                                                                                                                              
Select the way of assign the project root                                                                                                                                                                                                       
If cancelled, "/home/_/.vim/pack/bundle/start/phpactor" will be used.                                                                                                                                                                      
1: manual (default "/tmp")                                                                                                                                                                                                                      
Type number and <Enter> or click with mouse (empty cancels):                                                                                                                                                                                    
Error detected while processing /home/_/.vim/pack/bundle/start/phpactor/ftplugin/php/autocommands.vim:                                                                                                                                     
line    1:                                                                                                                                                                                                                                      
Interrupted                                                                                                                                                                                                                                     
Error detected while processing /home/_/.vim/pack/bundle/start/phpactor/ftplugin/php/commands.vim:                                                                                                                                         
line    1:                                                                                                                                                                                                                                      
Interrupted                                                                                                                                                                                                                                     
Error detected while processing /home/_/.vim/pack/bundle/start/phpactor/ftplugin/php/mappings.vim:                                                                                                                                         
line    1:                                                                                                                                                                                                                                      
Interrupted                                                                                                                                                                                                                                     
Error detected while processing /home/_/.vim/pack/bundle/opt/_/ftplugin/php.vim:                                                                                                                                                      
line    1:                                                                                                                                                                                                                                      
Interrupted                                                                                                                                                                                                                                     
Error detected while processing /usr/share/vim/vim82/ftplugin/php.vim:                                                                                                                                                                          
line    1:                                                                                                                                                                                                                                      
Interrupted                                                                                                                                                                                                                                     
Interrupt: Press ENTER or type command to continue 
```
After moving calling back from ftplugin to autocommand, the error does not occur, but probably the rest of code is not executed (throwing anything gives no effect).

Reply:
I still have no idea how to correctly handle keyboard interruption (in viml).

Reply:
Outdated: `Interrupted` is an exception message, probably caught somewhere, so it did not look like an exception at first glance ;)